File: swift/Sources/benchmarks/GraphPathBenchmark.swift
import Foundation

class GraphPathBenchmark: BenchmarkProtocol {
    final class Graph {
        let vertices: Int
        let adj: [[Int]]

        init(vertices: Int, components: Int = 10) {
            self.vertices = vertices

            var tempAdj = Array(repeating: [Int](), count: vertices)
            let componentSize = vertices / components

            for c in 0..<components {
                let startIdx = c * componentSize
                var endIdx = (c + 1) * componentSize
                if c == components - 1 {
                    endIdx = vertices
                }

                for i in (startIdx + 1)..<endIdx {
                    let parent = startIdx + Helper.nextInt(max: i - startIdx)
                    tempAdj[i].append(parent)
                    tempAdj[parent].append(i)
                }

                for _ in 0..<(componentSize * 2) {
                    let u = startIdx + Helper.nextInt(max: endIdx - startIdx)
                    let v = startIdx + Helper.nextInt(max: endIdx - startIdx)
                    if u != v {
                        tempAdj[u].append(v)
                        tempAdj[v].append(u)
                    }
                }
            }

            self.adj = tempAdj
        }

        func generateRandom() {

        }
    }

    var graph: Graph!
    var pairs: [(Int, Int)] = []
    private var nPairs: Int64 = 0
    private var resultVal: UInt32 = 0

    init() {

    }

    func prepare() {
        if nPairs == 0 {
            nPairs = configValue("pairs") ?? 0
            let vertices = Int(configValue("vertices") ?? 0)
            let comps = max(10, vertices / 10_000)
            graph = Graph(vertices: vertices, components: comps)
            pairs = generatePairs(n: Int(nPairs))
        }
    }

    private func generatePairs(n: Int) -> [(Int, Int)] {
        var pairs = [(Int, Int)]()
        pairs.reserveCapacity(n)
        let componentSize = graph.vertices / 10

        for _ in 0..<n {
            if Helper.nextInt(max: 100) < 70 {
                let component = Helper.nextInt(max: 10)
                let start = component * componentSize + Helper.nextInt(max: componentSize)
                var end: Int
                repeat {
                    end = component * componentSize + Helper.nextInt(max: componentSize)
                } while end == start
                pairs.append((start, end))
            } else {
                let c1 = Helper.nextInt(max: 10)
                var c2 = Helper.nextInt(max: 10)
                while c2 == c1 {
                    c2 = Helper.nextInt(max: 10)
                }
                let start = c1 * componentSize + Helper.nextInt(max: componentSize)
                let end = c2 * componentSize + Helper.nextInt(max: componentSize)
                pairs.append((start, end))
            }
        }
        return pairs
    }

    func test() -> Int64 {
        return 0
    }

    func run(iterationId: Int) {
        resultVal &+= UInt32(test())
    }

    var checksum: UInt32 {
        return resultVal
    }

    var name: String { return "GraphPathBenchmark" }
}

File: csharp/GraphPathBenchmark.cs
public abstract class GraphPathBenchmark : Benchmark
{
    protected class Graph
    {
        public int Vertices { get; }
        public List<int>[] Adj { get; }
        private readonly int _components;

        public Graph(int vertices, int components = 10)
        {
            Vertices = vertices;
            _components = components;
            Adj = new List<int>[vertices];
            for (int i = 0; i < vertices; i++) Adj[i] = new List<int>();
        }

        public void AddEdge(int u, int v)
        {
            Adj[u].Add(v);
            Adj[v].Add(u);
        }

        public void GenerateRandom()
        {
            int componentSize = Vertices / _components;

            for (int c = 0; c < _components; c++)
            {
                int startIdx = c * componentSize;
                int endIdx = (c + 1) * componentSize;
                if (c == _components - 1) endIdx = Vertices;

                for (int i = startIdx + 1; i < endIdx; i++)
                {
                    int parent = startIdx + Helper.NextInt(i - startIdx);
                    AddEdge(i, parent);
                }

                for (int i = 0; i < componentSize * 2; i++)
                {
                    int u = startIdx + Helper.NextInt(endIdx - startIdx);
                    int v = startIdx + Helper.NextInt(endIdx - startIdx);
                    if (u != v) AddEdge(u, v);
                }
            }
        }
    }

    protected Graph _graph;
    protected List<(int, int)> _pairs;
    protected uint _result;
    protected long _nPairs;

    protected GraphPathBenchmark()
    {
        _result = 0;
        _nPairs = ConfigVal("pairs");
    }

    protected List<(int, int)> GeneratePairs(int n)
    {
        var pairs = new List<(int, int)>();
        int componentSize = _graph.Vertices / 10;

        for (int i = 0; i < n; i++)
        {
            if (Helper.NextInt(100) < 70)
            {
                int component = Helper.NextInt(10);
                int start = component * componentSize + Helper.NextInt(componentSize);
                int end;
                do { end = component * componentSize + Helper.NextInt(componentSize); } while (end == start);
                pairs.Add((start, end));
            }
            else
            {
                int c1 = Helper.NextInt(10);
                int c2;
                do { c2 = Helper.NextInt(10); } while (c2 == c1);

                int start = c1 * componentSize + Helper.NextInt(componentSize);
                int end = c2 * componentSize + Helper.NextInt(componentSize);
                pairs.Add((start, end));
            }
        }

        return pairs;
    }

    public override void Prepare()
    {
        int vertices = (int)ConfigVal("vertices");
        int comps = Math.Max(10, vertices / 10000);
        _graph = new Graph(vertices, comps);
        _graph.GenerateRandom();
        _pairs = GeneratePairs((int)_nPairs);
    }

    protected abstract long Test();

    public override void Run(long IterationId) => _result += (uint)Test();

    public override uint Checksum => _result;
}